perf/xtm.cc

```diff
@@ -0,0 +1,332 @@
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include "xtm/xtm_api.h"
+#include "../test/unit.h"
+/** Conflict name with same macro in benchmark.h */
+#undef is
+
+#include <pthread.h>
+
+#include <benchmark/benchmark.h>
+
+enum {
+	/** Maximum message count in xtm queue. */
+	XTM_TEST_QUEUE_SIZE = 64 * 1024,
+	/**
+	 * Maximum number of messages sent without notification.
+	 * Upper bound for the test parameter.
+	 */
+	BATCH_COUNT_MAX = 1024,
+	/** Count of messages in test */
+	TEST_MSG_COUNT = 1024 * 1024,
+};
+
+struct xtm_msg {
+	unsigned number;
+};
+
+/**
+ * Global pointer to xtm queue.
+ */
+static struct xtm_queue *xtm_queue;
+/** Consumer thread id */
+static pthread_t consumer_thread;
+/**
+ * Global message preallocated pointers array, we used it,
+ * because we want to test performance of xtm, not of malloc.
+ */
+static struct xtm_msg *xtm_msg_arr[TEST_MSG_COUNT];
+
+static int
+wait_for_fd(int fd)
+{
+	fd_set readset;
+	FD_ZERO(&readset);
+	FD_SET(fd, &readset);
+	return select(fd + 1, &readset, NULL, NULL, NULL);
+}
+
+static void
+consumer_msg_func(void *arg)
+{
+	struct xtm_msg *msg = (struct xtm_msg *)arg;
+	xtm_msg_arr[msg->number] = NULL;
+	free(msg);
+}
+
+static void *
+consumer_thread_push_and_invoke_fun(MAYBE_UNUSED void *arg)
+{
+	unsigned invoked = 0;
+	int fd = xtm_queue_consumer_fd(xtm_queue);
+	bool is_need_to_notify_again = false;
+
+	while (invoked < TEST_MSG_COUNT) {
+		fail_unless(wait_for_fd(fd) > 0);
+		fail_unless(xtm_queue_consume(fd) == 0);
+		int rc = xtm_queue_invoke_funs_all(xtm_queue);
+		if (rc < 0) {
+			/*
+			 * If an error occurs, function returns -(count of
+			 * invoked functions + 1). So we can calsulate count
+			 * of invoked functions as -(rc + 1).
+			 */
+			rc = -(rc + 1);
+			is_need_to_notify_again = true;
+		}
+		invoked += rc;
+		/* Try to notify producer again, if it's fails - panic */
+		if (is_need_to_notify_again) {
+			fail_unless(xtm_queue_notify_producer(xtm_queue) == 0);
+			is_need_to_notify_again = false;
+		}
+	}
+	return (void *)NULL;
+}
+
+static void *
+consumer_thread_push_and_pop_ptr(MAYBE_UNUSED void *arg)
+{
+	unsigned received = 0;
+	int fd = xtm_queue_consumer_fd(xtm_queue);
+	bool is_need_to_notify_again = false;
+
+	while (received < TEST_MSG_COUNT) {
+		fail_unless(wait_for_fd(fd) > 0);
+		fail_unless(xtm_queue_consume(fd) == 0);
+		unsigned count = xtm_queue_count(xtm_queue);
+		unsigned cnt = 0;
+		while (cnt < count) {
+			void *ptr_array[BATCH_COUNT_MAX];
+			unsigned popcnt = (count - cnt >= BATCH_COUNT_MAX ?
```

Comment:
Why can't you always pass `BATCH_COUNT_MAX` and use the returned count?

Reply:
Fixed